src/qmc_states.jl

```diff
@@ -0,0 +1,248 @@
+"""
+struct SpectralState{
+    N,
+    NS<:NTuple{N,SingleState},
+    SS<:SpectralStrategy{N}
+} <: AbstractVector{SingleState}
+    single_states::NS # Tuple of SingleState
+    spectral_strategy::SS # SpectralStrategy
+    id::String # id is appended to column names
+end
+function SpectralState(t::Tuple, ss::SpectralStrategy, id="")
+    return SpectralState(t, ss, id)
+end
+num_spectral_states(::SpectralState{N}) where {N} = N
+
+Base.size(s::SpectralState) = (num_spectral_states(s),)
+Base.getindex(s::SpectralState, i::Int) = s.single_states[i]
+
+function Base.show(io::IO, s::SpectralState)
+    print(io, "SpectralState")
+    print(io, " with ", num_spectral_states(s), " spectral states")
+    print(io, "\n    spectral_strategy: ", s.spectral_strategy)
+    for (i, r) in enumerate(s.single_states)
+        print(io, "\n      $i: ", r)
+    end
+end
+
+function state_vectors(state::SpectralState)
+    return mapreduce(state_vectors, vcat, state.single_states)
+end
+
+
+"""
+    _n_walkers(v, shift_strategy)
+Returns an estimate of the expected number of walkers as an integer.
+"""
+function _n_walkers(v, shift_strategy)
+    n = if hasfield(typeof(shift_strategy), :targetwalkers)
+        shift_strategy.targetwalkers
+    else # e.g. for LogUpdate()
+        walkernumber(v)
+    end
+    return ceil(Int, max(real(n), imag(n)))
+end
+
+"""
+    ReplicaState <: AbstractMatrix{SingleState}
+
+Holds information about multiple replicas of [`SpectralState`](@ref)s.
+
+Indexing the `ReplicaState` `state[i, j]` returns a `SingleState` from the `i`th replica
+and `j`th spectral state.
+
+See also [`ReplicaStrategy`](@ref), [`SpectralState`](@ref)s, [`SingleState`](@ref),
+[`PMCSimulation`](@ref).
+"""
+struct ReplicaState{
+    N, # number of replicas
+    S, # number of spectral states
+    R<:NTuple{N,<:SpectralState{S}},
+    RS<:ReportingStrategy,
+    RRS<:ReplicaStrategy,
+    PS<:NTuple{<:Any,PostStepStrategy},
+} <: AbstractMatrix{SingleState}
+    spectral_states::R
+    maxlength::Ref{Int}
+    step::Ref{Int}
+    simulation_plan::SimulationPlan
+    reporting_strategy::RS
+    post_step_strategy::PS
+    replica_strategy::RRS
+end
+
+
+num_replicas(::ReplicaState{N}) where {N} = N
+num_spectral_states(::ReplicaState{<:Any, S}) where {S} = S
+
+function Base.show(io::IO, st::ReplicaState)
+    print(io, "ReplicaState")
+    print(io, " with ", num_replicas(st), " replicas  and ")
+    print(io, num_spectral_states(st), " spectral states")
+    print(io, "\n  H:    ", first(st).hamiltonian)
+    print(io, "\n  step: ", st.step[], " / ", st.simulation_plan.last_step)
+    print(io, "\n  replicas: ")
+    for (i, r) in enumerate(st.spectral_states)
+        print(io, "\n    $i: ", r)
+    end
+end
+
+Base.size(st::ReplicaState) = (num_replicas(st), num_spectral_states(st))
+Base.getindex(st::ReplicaState, i::Int, j::Int) = st.spectral_states[i].single_states[j]
+Base.IndexStyle(::Type{<:ReplicaState}) = IndexCartesian()
+
+"""
```

Comment:
I think I'd prefer if this struct was unexported and would be accessed through an exported `state_vectors`.